### PR TITLE
util/generic: add a generic version of sync.Map

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/35983

Problem Summary:

It is used for managing the "backend contexts" and "engine information" in the implementation of the lightning backfill process.

### What is changed and how it works?

This PR adds a generic version of `sync.Map`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/lightning/sync_map.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+func NewSyncMap[K comparable, V any](capacity int) SyncMap[K, V] {
+	return SyncMap[K, V]{
+		item: make(map[K]V, capacity),
+	}
+}
+
+// Store stores a value.
+func (m *SyncMap[K, V]) Store(key K, value V) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
```
Comment (xhebox) [inline ddl/lightning/sync_map.go:35]: Maybe we can avoid using defer for some functions.

File: ddl/lightning/sync_map.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lightning
+
+import "sync"
+
+// SyncMap is the generic version of the sync.Map.
+type SyncMap[K comparable, V any] struct {
```
Comment (hawkingrei) [inline ddl/lightning/sync_map.go:20]: It can be put into ```util```. I think it is a good base library.

File: ddl/lightning/sync_map.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	defer m.mu.RUnlock()
+	val, exist := m.item[key]
+	if !exist {
+		return *new(V), exist
+	}
+	return val, exist
+}
+
+// Drop drops a value.
+func (m *SyncMap[K, V]) Drop(key K) {
```
Comment (hawkingrei) [inline ddl/lightning/sync_map.go:48]: It can be the same method to use as the ```Sync.Map```.

SyncMap.Drop is the same as ```Sync.Map.Delete```.

**Review by xhebox — APPROVED** [review]

File: util/generic/sync_map.go [hunk 4]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package lightning
```
Comment (hawkingrei) [inline util/generic/sync_map.go]: We can call it ```util/map```. We have ```util/set```. it will be added some generic version of ```set``` in the future.

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**